### ISSUE TITLE
enhancement: daemonized API service

### ISSUE THREAD

**Issue by aioue** [smira/aptly#264]
aptly should install and start its API service automatically in the background on install and boot like a standard linux service (nginx etc).

Configuration should be stored in /etc/ and working files (DB etc) in its own directory.

[![Bountysource](https://www.bountysource.com/badge/issue?issue_id=21733238)](https://www.bountysource.com/issues/21733238-enhancement-daemonized-api-service?utm_source=21733238&utm_medium=shield&utm_campaign=ISSUE_BADGE)

**Comment by AlekSi** [comment]
It should be easy with https://github.com/fhd/init-script-template

**Comment by oCCtet** [comment]
Automatically... not until #234 is resolved, as api serve apparently makes cli use impossible.

I guess one possibility would be to have an option in /etc/default/aptly, which - if set to a value like `AUTO_API_SERVE=1` - would cause aptly initscript to start the service on boot.

**Comment by aioue** [comment]
As long as the service installs and runs and acts a lot more like a 'regular' service, how it's accomplished is broadly irrelevant.

I'd expect to see maybe an 'aptly' user created by the package, an `/etc/aptly` directory for config, and the DB data kept in it's own directory, something like `/usr/local/aptly`.

#234 isn't a blocker. You have to tell a service like haproxy to be enabled in the config before it'll run automatically. When the CLI eventually moves to calling the API under the covers, then even that requirement will go away.

**Comment by oCCtet** [comment]
Originally aptly was not limited to one-instance-per-machine; for example, each user could maintain their own package repositories (think PPAs...). The proposal of automatically installing aptly as a machine-wide service strongly suggests a severely limited use case.

IMHO making the CLI use API service as the backend is a terrible idea. It should be the other way around: higher-level interfaces should use lower-level functions, i.e. HTTP interface should use command-line utilities (or library calls if such were available) to perform its tasks.

**Comment by aioue** [comment]
Running as a service is no different from multiple people trying to bind to the same default port using `aptly api serve`. I'm only repeating smira's ideas in #234 re: CLI calling API since he's the owner of the project and ultimately decides its direction.

@AlekSi if you can get it functioning as it is, but working as a persistent service via some init scripts, the bounty is yours.

**Comment by aioue** [comment]
This looks promising:
"A daemon package for use with Go (golang) services with no dependencies."
https://github.com/takama/daemon

**Comment by igorastds** [comment]
I've checked https://github.com/takama/daemon/blob/master/daemon_linux_systemv.go#L202 - here it can be seen that takama/daemon basically installs quite strange init script (which is using `/etc/rc.d/init.d/functions`, when pure Debian is all up for /lib/lsb/init-function )..

I think I can implement the init script thingamabob, what distros should be supported by init scripts? wheezy/sid, ubuntu, others?

**Comment by aioue** [comment]
Your suggested wheezy/sid and Ubuntu sound good. Both Debian and Ubuntu are moving towards systemd, but Ubuntu 14.04 LTS support is going to run to 2019 so it's a worthwhile target.

**Comment by igorastds** [comment]
@aioue ok, I'm starting it @ bountysource..

**Comment by avongluck-r1soft** [comment]
Just create an upstart job... systemd should be just as easy.

```
# cat /etc/init/aptly.conf

# Aptly - Deb package management
#
# Manage all the debs!

description	"Aptly deb repo"
start on runlevel [2345]

setuid "aptlyuser"
setgid "aptlyuser"
chdir "/home/aptlyuser"

exec /usr/local/bin/aptly api serve -listen "127.0.0.1:8080"
```

**Comment by aioue** [comment]
That's only half the story. Config is currently read from /etc/ or ~, but a
default install will create everything in the user's home dir rather than,
say config in `/etc` and data (database etc) in `/opt` (or similar).

http://www.aptly.info/doc/configuration/

On 14 August 2015 at 15:32, Alexander von Gluck IV <notifications@github.com
> wrote:

> Just create an upstart job... systemd should be just as easy.
>
> # cat /etc/init/aptly.conf
>
> # Aptly - Deb package management
> #
> # Manage all the debs!
>
> description "Aptly deb repo"
> start on runlevel [2345]
>
> setuid "aptlyuser"
> setgid "aptlyuser"
> chdir "/home/aptlyuser"
>
> exec /usr/local/bin/aptly api serve -listen "127.0.0.1:8080"
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/smira/aptly/issues/264#issuecomment-131124977>.
>



-- 
+44 7786 972825

**Comment by aioue** [comment]
@igorastds any news? Bountysource thinks you're still working on it. If you've stopped maybe un-link yourself so people are encouraged to claim the bounty.

**Comment by fabrepe** [comment]
I began to work on this task. Could you clarify a point :
- is a package (eg. aptly-daemon) that embed daemonizing and configuration scripts a solution ? It would has a dependency on aptly packages.

**Comment by aioue** [comment]
@fabrepe I would expect aptly to work a lot like (for example) an nginx install.

1. User installs aptly using their package manager
 - package install creates `/etc/aptly/` directory skeleton and default config file
 - installs aptly binaries in a standard location, e.g. `/usr/sbin/`
 - installs scripts to daemonize the existing HTTP REST endpoint so that it starts on boot when enabled

1. When first run, aptly installs it's database, package and published repository storage into a sensible shared directory e.g. `/opt/aptly/`

1. Backwards compatibility can be retained by keeping [the existing config file priorities](http://www.aptly.info/doc/configuration/). Only the default locations for the binary and data files would be different, and it would turn aptly into a fully fledged system service.

**Comment by fabrepe** [comment]
@aioue again some questions :

1. You said : "Backwards compatibility can be retained by keeping the existing config file priorities."
Do you mean that users used to run aptly as root ? therefore they need to load their configuration files (~/aptly.conf, or /etc/aptly.conf)

2.  In the end of the task, do you want an aptly package ?

Let me pick up some details :

1.  Standard location may stay at /usr/bin/aptly, allowing common users to run in parallel their own instance of aptly (on another port). Security issues are pruned by creating an aptlyuser without belonging to root or sudo group and running the daemon as this system user. In that case, no rights escalation can occur by replacing the aptly binary.

2.   According to the [Filesystem Hierarchy Standard](https://wiki.debian.org/FilesystemHierarchyStandard), I would rather put database at /var/local/aptly

**Comment by kumy** [comment]
> According to the Filesystem Hierarchy Standard, I would rather put database at /var/local/aptly

A Debian package should never put files in `local` directories.
See §9.1.2 https://www.debian.org/doc/debian-policy/ch-opersys.html

**Comment by fabrepe** [comment]
@kumy Section 9.1.2 is only  refering to /usr/local. No restrictions for /var.

**Comment by kumy** [comment]
@fabrepe right :)
This has been [removed from Lintian checks](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=119414) in 2002!

**Comment by aioue** [comment]
@fabrepe 

1. You're right maybe aptly install package should also create an `aptly` user instead. (--as you have suggested)

2. At the end yes, a new aptly package would be the goal.

"Standard location may stay at /usr/bin/aptly, allowing common users to run in parallel their own instance of aptly (on another port). Security issues are pruned by creating an aptly user without belonging to root or sudo group and running the daemon as this system user. In that case, no rights escalation can occur by replacing the aptly binary."

I agree with everything you have said, and also the database files going to `/var/aptly/` instead.

**Comment by fabrepe** [comment]
ok, I may have finished the work. The only issue is that at installation, no repository are published. As a consequence, aptly cannot start.

I have to do a push request (however i'm not familiar with such commands). 
Please let me know if you face some issues.

**Comment by aioue** [comment]
@fabrepe raise a PR against smira's project (this repo)

**Comment by smira** [comment]
Debian packaging script are kept in Debian repo, authored by Debian maintainer. I don't see why they should be pushed to this repo. See #312

**Comment by fabrepe** [comment]
Ok, sorry for the delay. 
I come into connection with the package maintainer to integrate changes.
However, to my mind, some scripts will stille have to be pushed onto this repo.

**Comment by jmihalich** [comment]
Hi, i'm desperately waiting for this feature, and a little confused by the status of this.  Is this going to be merged into the master aptly branch, and will a new version be released shortly thereafter that has this stuff built in, so it installs itself as a service when you install the main aptly deb package?

**Comment by Pryz** [comment]
Any new on this ?

**Comment by sdumetz** [comment]
Having it pushed to debian/Ubuntu repo is the way to go but it will take time. Meanwhile, users that really need this should : 

- create an **aptly** user with home in `/var/aptly` 
- create an init script in `/etc/systemd/system/aptly.service` (for systemd users)

More people using it means more chance to know if anything is going wrong with this approach. Meanwhile, the package maintainers can include those simple steps in their future versions. 

I don't think there is any work to do in aptly for this since (afaik) upstream are **not** supposed to create users and services in their installs unless necessary.

My systemd script for reference (nothing fancy here) : 

```
[Unit]
Description=custom aptly REST API service

[Service]
User=aptly
Group=aptly
WorkingDirectory=/var/aptly
ExecStart=/usr/bin/aptly api serve -listen ":8081"

[Install]
WantedBy=multi-user.target
```

**Comment by jola5** [comment]
I second @smira on this, a service script is part of the distribution package (refer to https://packages.qa.debian.org/a/aptly.html maintained by https://qa.debian.org/developer.php?login=seb@debian.org). And as pointed out by @sdumetz this is pretty easy to achieve with a systemd unit file. 

This request is nearly two years old without any improvement.  I suggest to close this with `wontfix`.

**Comment by smira** [comment]
I agree, thanks
